### PR TITLE
Command registry

### PR DESCRIPTION
**Pull request by kreneskyp**
This PR implements a command registry enabling custom commands added at runtime.

## Changes:
- adds` Command` and `CommandRegistry` classes for defining and registering commands
- adds `command` decorator to support existing functions
- CommandRegistry intialized at startup loading commands from `/scripts/`
- main loop now loads commands from registry instance
- file operation commands now accept `file` instead of `filename` to match prompt args.

## TODO:
- [ ] The prompt is still hardcoded. Need to turn into a template
- [ ] Need to handle the remaining edgecase commands
- [ ] Does order of plugins matter in prompt?
- [ ] consolidate into `scripts/commands.py`
- [ ] tie into configuration somehow?
- [ ] unit tests

### CODE REVIEW

**Comment by dominikdassow** [comment]
Great work!

I think the next, complementary step would be a "command repository" that can be used to individually enable certain commands, and also load additional/external/plugin commands. This command repository would then be listing the commands in the prompt.txt.

@ryanpeach's amazing PR #146 would certainly also benefit from such a command repository.
Would also make easier for other commands, like #437.

Separate PR for that or is that maybe more related here?

**Comment by ryanpeach** [comment]
I noticed this is 100% needed!

I would like a central place for all Prompt related constants.

Also I would like it if in the config yamls I call "personalities" in my PR, if you could whitelist or blacklist commands. Some agents just don't need the freedom to write/read files or run code, and that makes their action space way too complex for them.

**Comment by ryanpeach** [comment]
Hey so this pr creates an auto_gpt folder.

My pr makes scripts an module

This https://github.com/Torantulino/Auto-GPT/pull/329 PR renames the scripts folder to autogpt and changes the import.

This is a critical Basic Code Quality issue @Torantulino https://github.com/Torantulino/Auto-GPT/issues/137

**Comment by kreneskyp** [comment]
> Hey so this pr creates an auto_gpt folder to handle the lack of a module.

Yeah i started moving it to a namespace, which is badly needed, but didn't want to do a bulk move in this branch.

**Comment by ryanpeach** [comment]
I recommend consolodating all "turning the code into a module" PR's I mentioned into this one. I'm working to reconcile them all.

https://github.com/Torantulino/Auto-GPT/pull/142

Could you rename your new folder `autogpt` instead of `auto_gpt`?

**Comment by bnuttall** [comment]
Big fan of this

We also need to figure out how to do better hinting so the AI can use tools more effectively.

I did a little Wolfram Alpha tool last night and someone pointed me to the official OpenAI plug in. It's got a really good hint which I guess increases the hit rate of GPT-4 correctly using the solver (though it uses a ton of tokens to do it). With mine you kind of have to lead your AI by the nose to use the solver the first couple of times, but once it does then it will start using it correctly and consistently. A better hint could make this much closer to one shot.

(Also another vote in favor of persistent memory, I've taught my persistent AIs to use solvers and they don't forget now between sessions.)

**Comment by Torantulino** [comment]
Let's get this in after #142.

> Could you rename your new folder autogpt instead of auto_gpt?

@kreneskyp 
Can we get this resolved so that they're not solving the same problem in a different way?

File: scripts/commands.py [hunk 1]
```diff
@@ -52,64 +47,27 @@ def get_command(response):
...
-            return ai.write_tests(arguments["code"], arguments.get("focus"))
-        elif command_name == "execute_python_file":  # Add this command
-            return execute_python_file(arguments["file"])
-        elif command_name == "generate_image":
-            return generate_image(arguments["prompt"])
         elif command_name == "task_complete":
             shutdown()
         else:
             return f"Unknown command {command_name}"
-    # All errors, return "Error: + error message"
```
Comment (nponeccop) [inline scripts/commands.py:112]: Avoid unrelated changes. The scope is already large.

File: scripts/file_operations.py [hunk 2]
```diff
@@ -1,5 +1,7 @@
 import os
 import os.path
+from auto_gpt.commands import command
+
```
Comment (nponeccop) [inline scripts/file_operations.py:5]: Extra blank line.

**Review by nponeccop — APPROVED** [review]
Minor tweaks (IMO the PR can be merged as is if it works).

**Comment by nponeccop** [comment]
@kreneskyp Implement the suggestions and resolve the conflicts.

**Comment by kreneskyp** [comment]
[nponeccop](https://github.com/nponeccop) Sorry got busy with other stuff and didn't have time to finish this. Feel free to take this over and get it merged.

**Comment by BillSchumacher** [comment]
> [nponeccop](https://github.com/nponeccop) Sorry got busy with other stuff and didn't have time to finish this. Feel free to take this over and get it merged.

I'm going to pull these changes into the plugin support repo and get the conflicts resolved. Thanks for the great work!

**Comment by BillSchumacher** [comment]
This is now merged into #757 I have resolved conflicts but still need to move some stuff over.
